hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/ratis/BaseRatisCommand.java

```diff
@@ -0,0 +1,132 @@
+import static org.apache.ratis.shell.cli.RaftUtils.buildRaftPeersFromStr;
+import static org.apache.ratis.shell.cli.RaftUtils.retrieveGroupInfoByGroupId;
+import static org.apache.ratis.shell.cli.RaftUtils.retrieveRemoteGroupId;
+
+/**
+ * The base class for all the ratis command.
+ */
+public class BaseRatisCommand {
+  private RaftGroup raftGroup;
+  private GroupInfoReply groupInfoReply;
+  private OzoneConfiguration config;
+  private String omServiceId;
+
+  public void run(String peers, String groupid, String omServiceId) throws Exception {
+    this.config = new OzoneConfiguration();
+    this.omServiceId = omServiceId;
+    List<RaftPeer> peerList = buildRaftPeersFromStr(peers);
+    RaftGroupId raftGroupIdFromConfig = buildRaftGroupIdFromStr(groupid);
+    raftGroup = RaftGroup.valueOf(raftGroupIdFromConfig, peerList);
+
+    try (RaftClient client = buildRaftClient(raftGroup)) {
+      RaftGroupId remoteGroupId = retrieveRemoteGroupId(raftGroupIdFromConfig, peerList, client, System.out);
+      groupInfoReply = retrieveGroupInfoByGroupId(remoteGroupId, peerList, client, System.out);
+      raftGroup = groupInfoReply.getGroup();
+    }
+
+  }
+
+  protected RaftClient buildRaftClient(RaftGroup raftGroup) throws Exception {
+    GrpcTlsConfig tlsConfig = null;
+    if (config.getBoolean(OZONE_SECURITY_ENABLED_KEY, OZONE_SECURITY_ENABLED_DEFAULT) &&
+        config.getBoolean(HDDS_GRPC_TLS_ENABLED, HDDS_GRPC_TLS_ENABLED_DEFAULT)) {
+      tlsConfig = createGrpcTlsConf(omServiceId);
+    }
+
+    return RatisHelper.newRaftClient(
+        SupportedRpcType.GRPC, null, raftGroup,
+        RatisHelper.createRetryPolicy(config), tlsConfig, config);
+  }
+
+  public RaftGroup getRaftGroup() {
+    return raftGroup;
+  }
+
+  public GroupInfoReply getGroupInfoReply() {
+    return groupInfoReply;
+  }
+
+  /**
+   * Get the leader id.
+   *
+   * @param roleInfo the role info
+   * @return the leader id
+   */
+  protected RaftPeerProto getLeader(RoleInfoProto roleInfo) {
+    if (roleInfo == null) {
+      return null;
+    }
+    if (roleInfo.getRole() == RaftPeerRole.LEADER) {
+      return roleInfo.getSelf();
+    }
+    FollowerInfoProto followerInfo = roleInfo.getFollowerInfo();
+    if (followerInfo == null) {
+      return null;
+    }
+    return followerInfo.getLeaderInfo().getId();
+  }
+
+  private GrpcTlsConfig createGrpcTlsConf(String omServiceId) throws Exception {
+    OzoneClient certClient = OzoneClientFactory.getRpcClient(omServiceId,
+        config);
+
+    ServiceInfoEx serviceInfoEx = certClient
+        .getObjectStore()
+        .getClientProxy()
+        .getOzoneManagerClient()
+        .getServiceInfo();
+
+    CACertificateProvider remoteCAProvider =
+        serviceInfoEx::provideCACerts;
+    ClientTrustManager trustManager = new ClientTrustManager(remoteCAProvider, serviceInfoEx);
+
+    return RatisHelper.createTlsClientConfig(new SecurityConfig(config), trustManager);
```

Comment:
At this point we also have a potential problem...

As I understand this PR, the goal is to provide CLI commands that an admin can run and via which the admin can issue commands to a Ratis Group directly from the CLI.

If this understanding of mine is correct, then we may run into a problem with mutual TLS authentication that happens between the raft peers, because the createTlsClientConfig call creates a client side config that does not apply mTLS because the client does not have its own certificate that our internal PKI system trusts, and also it does not have a way to acquire such a certificate from the internal PKI system.

So unless the client here talks to a ratis client port that does not require mTLS but uses other authentication mechanisms like Kerberos, then this approach ain't gonna fly and honestly in this case we need to figure out if we can and how we can allow the client to get its own certificate to use mTLS with the raft server side. (Also we might need to offload this from SCM if these calls are something clients would require potentially often, as we can not afford to generate a lot of certificate on the SCM side as it is a costly operation.) Not to mention the complexity of ensuring that only valid users can grab a cert from the system.
OTOH, if these raft servers do not use mTLS then how they authenticate and identify each others?

Reply:
I think currently the client side does not require mTLS authentication (per comment in https://github.com/apache/ozone/blob/master/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java#L421-L422). But I could be wrong, does any other folk have additional insights or recommendations regarding the authentication and security aspects of this problem? (cc. @szetszwo @ChenSammi @Galsza

Reply:
I will let @szetszwo to chime in on this one... if we need mTLS the problem is there, if the API itself does not require authentication, then we should be good to go.

However, I think, if there is no authentication then we need to be sure that there are no administrative commands that can be triggered via the Ratis client and we also need to be sure that there is no sensitive information that are emitted in the response.